Add support for DNF types

IonBazan opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no (not yet)
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

Basic support for DNF types (as per #142)

Code review welcome, I'd like to know if this is the right direction. Collaboration is welcome too 🙏🏻 .

IonBazan committed:
Add basic DNF type objects

Signed-off-by: Ion Bazan <ion.bazan@gmail.com>

Ocramius commented on src/Generator/TypeGenerator/AtomicType.php:
```diff
@@ -128,6 +128,11 @@ public function fullyQualifiedName(): string
             : '\\' . $this->type;
     }
 
+    public function __toString(): string
+    {
+        return $this->fullyQualifiedName();
+    }
```
Let's use `->toString()` please :) 

Rationale: https://github.com/ShittySoft/symfony-live-berlin-2018-doctrine-tutorial/pull/3#issuecomment-460441229

Ocramius commented on src/Generator/TypeGenerator/BaseComplexType.php:
```diff
@@ -0,0 +1,23 @@
+
+namespace Laminas\Code\Generator\TypeGenerator;
+
+abstract class BaseComplexType implements Type
```
`/** @internal */` for anything we didn't build explicitly for external systems

Ocramius replied:
s/`BaseComplexType`/`CompositeType`? 

We already know it's a "base" type from the `abstract` keyword

Ocramius commented on src/Generator/TypeGenerator/BaseComplexType.php:
```diff
@@ -0,0 +1,23 @@
+    /**
+     * @param AtomicType[]|IntersectionType[] $types
+     */
+    final protected function __construct(protected array $types)
```
Uhh, let's rather please just use union types, instead of an abstract class with `protected` fields.

Really messy to maintain this later on otherwise :)

Ocramius commented on test/Generator/TypeGenerator/UnionTypeTest.php:
```diff
@@ -0,0 +1,27 @@
+    {
+        yield ['A|B', '\\A|\\B'];
+        yield ['foo|bar|baz|null', '\\foo|\\bar|\\baz|null'];
+        yield ['(foo&bar)|baz|null', '(\\foo&\\bar)|\\baz|null'];
```
We need `(foo&bar)|(baz&taz)|(tar&war&waz)` or such to show that unions are not just across 2 types

Ocramius reviewed (changes requested):
Let's get psalm happy, before continuing this :D

IonBazan committed:
Refactor

Signed-off-by: Ion Bazan <ion.bazan@gmail.com>

IonBazan committed:
Add DNF test

Signed-off-by: Ion Bazan <ion.bazan@gmail.com>

IonBazan commented:
The tests are passing on both PHP 8.1 and 8.2. I added some more test cases too.

Not sure if you want to keep `CompositeTypeTest` as it overlaps with what `TypeGeneratorTest` checks but it's definitely much easier to debug things having a separate test.

There is a lot of mess and code duplication in `toString()` and `__toString()` methods - not quite sure how to handle that nicely but I'm open to suggestions. Technically I could move everything back to the `TypeGenerator` and allow storing `TypeGenerator[]` in `types` but I feel it would give it too many responsibilities. Besides, it doesn't have a common interface with `AtomicType` (and shouldn't).

Idk why is psalm crashing 🤷🏻‍♂️

IonBazan committed:
Add better validation for parenthesis

Signed-off-by: Ion Bazan <ion.bazan@gmail.com>

Ocramius commented on src/Generator/TypeGenerator/TypeInterface.php:
```diff
@@ -0,0 +1,15 @@
+
+use Stringable;
+
+interface TypeInterface extends Stringable
```
Let's keep this stuff `internal` too :D

Ocramius reviewed (commented):
Overall looking good from the tests: implementation will need a lot of careful reading / focus :) 

Otherwise nice work! 

I'll check it out locally when I can, and get it shipped.

Ocramius commented:
The crash in psalm is... novel :O 

I wonder if we should just try going `vimeo/psalm:^5.0.0-rc1`?

IonBazan committed:
Add more docs

Signed-off-by: Ion Bazan <ion.bazan@gmail.com>

IonBazan commented:
Psalm v5 crashes there too so no solution for that so far 😝

IonBazan commented on src/Generator/TypeGenerator.php:
```diff
@@ -136,30 +129,14 @@ public static function fromTypeString(string $type): self
-        $this->types          = $types;
-        $this->nullable       = $nullable;
-        $this->isIntersection = $isIntersection;
+        if ($nullable && $type instanceof AtomicType) {
+            $type->assertCanBeStandaloneNullable();
+        }
```
Should I keep it here or better move back to `fromString()`?

Ocramius replied:
Works here too :+1:

IonBazan committed:
fix psalm

Signed-off-by: Ion Bazan <ion.bazan@gmail.com>

IonBazan commented:
Well, Psalm is kinda fixed now - produces many more errors but at least the analysis completes. You may want to add them to baseline later.
![image](https://user-images.githubusercontent.com/1985514/204293391-61db88a2-84eb-49e2-bb65-98d34495e6ef.png)

Locked at `dev-master` as `v5.0.0-rc1` does not allow PHP 8.2.

Ocramius committed:
Upgrading to `vimeo/psalm:5.1.0`, which no longer crashes when inspecting DNF types

Disjunctive Normal Form types are hard, yo!

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius commented:
@IonBazan I'm picking this up: `vimeo/psalm:^5.1.0` fixed the issue with scanning DNF types, apparently :)

IonBazan commented:
Great, thanks! Let me know if any help is needed.

Ocramius committed:
Ignore test assets from static analysis

This stuff is a bunch of junk: on purpose.

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Ensuring `TypeGenerator\TypeInterface#__toString()` is pure and `non-empty-string`

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Making `CompositeType::fromString()` pure, as it doesn't lead to side effects

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
`CompositeTypes#getTypes()` always retrieves a `non-empty-list`

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Documenting `CompositeTypes#getSeparator()` so `non-empty-string` is always guaranteed to be generated

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Adding template types to `ClassReflection`

Since `vimeo/psalm:^5`, inheriting from a class that declares
generic types also does require generic types to be declared
while extending.

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Re-generating psalm baseline with `vimeo/psalm:^5` and provided config and type improvements

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius commented on src/Generator/TypeGenerator/AtomicType.php:
```diff
@@ -128,6 +128,11 @@ public function fullyQualifiedName(): string
             : '\\' . $this->type;
     }
 
+    public function __toString(): string
+    {
+        return $this->type;
+    }
```
`non-empty-string` needed

Ocramius commented on src/Generator/TypeGenerator/CompositeType.php:
```diff
@@ -0,0 +1,157 @@
+        return implode($this->getSeparator(), $typesAsStrings);
+    }
+
+    public function fullyQualifiedName(): string
```
Should be `non-empty-string`

Ocramius commented on src/Generator/TypeGenerator/CompositeType.php:
```diff
@@ -0,0 +1,157 @@
+        return $this->isIntersection ? self::INTERSECTION_SEPARATOR : self::UNION_SEPARATOR;
+    }
+
+    public function __toString(): string
```
Should be `non-empty-string`

Ocramius commented on src/Generator/TypeGenerator/CompositeType.php:
```diff
@@ -0,0 +1,157 @@
+ *
+ * @psalm-immutable
+ */
+final class CompositeType implements TypeInterface
+{
+    public const UNION_SEPARATOR        = '|';
+    public const INTERSECTION_SEPARATOR = '&';
```
Representing both union and intersection types as a single type is quite a mess: IMO not a solution we should roll with.

Ocramius reviewed (commented):
Still working on it - got stuck on paid work :P

Ocramius committed:
Ensuring string casts lead to `non-empty-string` in `TypeGenerator` internals

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Declaring a concept of `IntersectionType` containing sorted `AtomicType` instances

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Declaring a concept of `UnionType` containing sorted `AtomicType` and `IntersectionType` instances

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Enforce consistency of intersected types, simplified sorting accordingly

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Noting how `IntersectionType#$types` is sorted

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Better abstraction around how union and intersection types are validated internally

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Rewrote `CompositeType` and `TypeGenerator` to leverage the new union/intersection type abstractions

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Rewrote `TypeGenerator` to not convert to string and back for each type

This massively reduces the internal complexity
of the `TypeGenerator`, since we convert
`ReflectionType` symbols **directly** to our `@internal`
data structures, ready to be rendered.

Before this, we would cast the whole `ReflectionType` to a `string` that
fits our need, and then we would go back to parsing that string, with
a substantial overhead (especially considering the newly introduced
validation rules around DNF types).

Note that this introduces new Psalm violations that were added to `psalm-baseline.xml`,
but which are solved by my work @ https://github.com/vimeo/psalm/pull/8722

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Applied CS fixes (spacing, imported symbols, docblock structure)

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius commented:
@IonBazan as you can see, I ended up using a radically different representation of internal types, mostly so that it can be understood at runtime.

Shipping this now :)

Ocramius committed:
Disabling CS rule enforcing `Abstract` prefix on `abstract` classes: not useful in this codebase

Signed-off-by: Marco Pivetta <ocramius@gmail.com>
